File: account-service/src/main/java/com/wks/servicemarketplace/accountservice/adapters/auth/keycloak/KeycloakUser.java
package com.wks.servicemarketplace.accountservice.adapters.auth.keycloak;

import com.wks.servicemarketplace.accountservice.core.auth.User;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Value;

import javax.validation.constraints.NotNull;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

@Value
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Builder
class KeycloakUser implements User {
    @NotNull
    private final String uuid;
    private final String scope;
    @NotNull
    private final String email;
    private final boolean emailVerified;
    private final String name;
    @NotNull
    private final String preferredUsername;

    private final String givenName;
    private final String familyName;
    @NotNull
    private final List<String> roles;

    @Override
    public String getUuid() {
        return uuid;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getUsername() {
        return preferredUsername;
    }

    @Override
    public boolean hasRole(String role) {
        return Optional.ofNullable(roles)
                .orElseGet(Collections::emptyList)
                .stream()
                .anyMatch(it -> it.equalsIgnoreCase(role));
    }
}

File: job-service/src/main/kotlin/com/wks/servicesmarketplace/jobservice/adapters/auth/keycloak/KeycloakUser.kt
package com.wks.servicesmarketplace.jobservice.adapters.auth.keycloak

import com.wks.servicesmarketplace.jobservice.core.auth.Role
import com.wks.servicesmarketplace.jobservice.core.auth.User
import org.springframework.security.core.GrantedAuthority

data class KeycloakRole(override val name: String) : Role, GrantedAuthority {
    override fun getAuthority(): String {
        return name
    }
}

data class KeycloakUser(
        override val uuid: String,
        val scope: String? = null,
        val email: String,
        val emailVerified: Boolean,
        private val name: String,
        val preferredUsername: String,
        val givenName: String?,
        val familyName: String?,
        override val roles: List<KeycloakRole>
) : User {
    override fun getName() = this.name
    override val username = this.preferredUsername
}